src/helpers.rs

```diff
@@ -659,6 +664,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         }
     }
 
+    fn deref_operand_as(
+        &self,
+        op: &OpTy<'tcx, Provenance>,
+        layout: TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx, MPlaceTy<'tcx, Provenance>> {
+        let this = self.eval_context_ref();
+        let ptr = this.read_pointer(op)?;
+
+        let align = if this.machine.check_alignment != AlignmentCheck::None {
+            layout.align.abi
+        } else {
+            Align::ONE
+        };
+
+        this.check_ptr_access_align(ptr, layout.size, align, CheckInAllocMsg::DerefTest)?;
+
+        Ok(MPlaceTy::from_aligned_ptr(ptr, layout))
+    }
```

Comment:
This could be just `deref_operand_and_offset` with offset 0, right?

Reply:
That has an assert to make sure its inbounds of the original type, so it behaves differently from this, which does all checks based on `layout`

Reply:
Yeah... but I don't think there is a good reason for that.
what I don't like about this is that it essentially re-implements `deref_operand`. In particular we really shouldn't duplicate the `check_alignment` logic anywhere.

Reply:
Yeah, maybe this needs to live in rustc_const_eval so there can be shared code between this and deref_operand

Reply:
I have a simpler idea, I'll just push to your branch.

Reply:
Ah no that doesn't work. I'll make a rustc patch.